macros/datetime/date_spine.sql

```diff
@@ -60,7 +60,7 @@ filtered as (
 
     select *
     from all_periods
-    where date_{{datepart}} <= {{ end_date }}
+    where date_{{datepart}} <= {{ dbt_utils.safe_cast(end_date, dbt_utils.type_datetime()) }}
```

Comment:
I've spent more time than I'd like to admit looking at this line of SQL!

This gives me some amount of pause -- the `safe_cast` macro is intended to cast a _string_ to a more restrictive type, like a datetime. It is _not_ intended to be used with a target type (like a `datetime`) directly. That is:

```
-- This will work great (a string literal)
{{ dbt_utils.safe_cast("'2019-01-01'", dbt_utils.type_datetime()) }}

-- This will not work on Snowflake
-- eg: select try_cast('2019-01-01'::datetime as datetime)
{{ dbt_utils.safe_cast("'2019-01-01'::timestamp", dbt_utils.type_datetime()) }}
```


So, this PR represents a breaking change to `date_spine` on Snowflake at least -- it appears that BigQuery is a little bit more permissive with the types supplied to `safe_cast` though.

I can think of two alternatives:
1. force the caller of this macro to use `{{ dbt_utils.safe_cast(end_date, dbt_utils.type_datetime()) }}` in the macro which calls `date_spine` (we can better document this)
2. add an argument to `date_spine` which, if provided, will use `safe_cast` to cast  `start_date` and `end_date` parameters? (I think this probably isn't a very good approach).


What say you, @clausherther? You buy it?

Reply:
@drewbanin hmm, I see. So, this passed the tests b/c the integration test passes string literals and not date functions like in the usage docs. 

How do you feel about casting `field` in `safe_cast(field, type)` to a varchar in the Snowflake case? Or is that too much brute force?

```
{% macro snowflake__safe_cast(field, type) %}
    try_cast({{field}}::varchar as {{type}})
{% endmacro %}
```

Alternatively, how about **not** safe casting?
```
where date_{{datepart}} <= cast({{ end_date }} as {{ dbt_utils.type_datetime() }})
```

Reply:
Or, would something like this work? Cast the `end_date` only in the BQ case?

```
{# private, don't call from anywhere else #}
{%- macro prep_end_date_(end_date) -%}
  {{ adapter_macro('dbt_utils.prep_end_date_', end_date) }}
{%- endmacro -%}

{%- macro default__prep_end_date_(end_date) -%}
    {{ end_date }}
{%- endmacro -%}

{%- macro bigquery__prep_end_date_(end_date) -%}
    {{ dbt_utils.safe_cast(end_date, dbt_utils.type_datetime()) }}
{%- endmacro -%}
```

Used like:
```
    where date_{{datepart}} <= {{ dbt_utils.prep_end_date_(end_date) }}
```

Reply:
@drewbanin just wanted to check in to see what you thought re: the above.

Reply:
hey @clausherther - I was traveling over the past week - just getting caught up on PR comments now :)

I don't love the idea of casting to `varchar` in the Snowflake case, and I don't think auto-casting to datetime is a good idea either (ie. as shown in this example):
```
where date_{{datepart}} <= cast({{ end_date }} as {{ dbt_utils.type_datetime() }})
```

That last example you provided, `prep_end_date_` feels like the most reasonable option to me. This is really only necessary on BigQuery, so no need to introduce weirdness for other databases here. Maybe we can call it something explicit like `date_spine__parse_end_date` where the default implementation does nothing but the bq implementation casts to `datetime` as shown in your example.

I'll also say: I think there might be merit to better documenting the supported datatypes for the `date_spine` macro on each database, then requiring users to cast the `end_date` appropriately in their calling macro. That's certainly not as immediately convenient, but it could be a helpful constraint which simplifies the implementation of macros like `date_spine`.

Let me know what you think!

Reply:
Cool, makes sense. I like the "prep" macro option the best as well.
So, would it be best to just add this as a private macro inside of the date_spine macro file?

```

Reply:
Cool, makes sense. I like the "prep" macro option the best as well.
So, would it be best to just add this as a private macro inside of the date_spine macro file?
```
{# private, don't call from anywhere else #}
{%- macro date_spine__parse_end_date(end_date) -%}
  {{ adapter_macro('dbt_utils.prep_end_date_', end_date) }}
{%- endmacro -%}

{%- macro default__ date_spine__parse_end_date(end_date) -%}
    {{ end_date }}
{%- endmacro -%}

{%- macro bigquery__ date_spine__parse_end_date(end_date) -%}
    {{ dbt_utils.safe_cast(end_date, dbt_utils.type_datetime()) }}
{%- endmacro -%}
```

and then update the date_spine macro to use it?
```
where date_{{datepart}} <= {{ dbt_utils.date_spine__parse_end_date(end_date) }}
```

Reply:
This sounds great to me. I made one suggestion inline but will kick off the CI build when that's addressed!

Reply:
Ooops, sorry I missed that one, committed!